src/tests/krate.rs

```diff
@@ -618,6 +629,55 @@ fn versions() {
     );
 }
 
+#[test]
+fn recent_versions() {
+    use cargo_registry::schema::versions;
+
+    let (app, anon, user) = TestApp::init().with_user();
+    let user = user.as_model();
+    app.db(|conn| {
+        let _krate = CrateBuilder::new("foo_versions", user.id)
+            .version("0.5.1")
+            .version("1.0.0")
+            .expect_build(conn);
+
+        // Get the ids of the versions inserted
+        let v1: i32 = versions::table
+            .select(versions::id)
+            .filter(versions::dsl::num.eq("0.5.1"))
+            .first(conn)
+            .unwrap();
+        let v2: i32 = versions::table
+            .select(versions::id)
+            .filter(versions::dsl::num.eq("1.0.0"))
+            .first(conn)
+            .unwrap();
+
+        // Update the download count for some of these versions
+        VersionDownload::create_or_increment(v1, conn).unwrap();
+        VersionDownload::create_or_increment(v1, conn).unwrap();
+        VersionDownload::create_or_increment(v1, conn).unwrap();
+
+        VersionDownload::create_or_increment(v2, conn).unwrap();
+        VersionDownload::create_or_increment(v2, conn).unwrap();
```

Comment:
In general, I think this test setup is fine, but it's a bit verbose and a bit hard to match up with the assertions.

We have [a function on the crate builder](https://github.com/rust-lang/crates.io/blob/3e5dd28127284e64ff8e4b106ca5dc3472398e5a/src/tests/builders.rs#L212-L217) that sets recent downloads on a crate ([by adding downloads to an arbitrary version of the crate](https://github.com/rust-lang/crates.io/blob/3e5dd28127284e64ff8e4b106ca5dc3472398e5a/src/tests/builders.rs#L260-L270)), I think we could do something similar on the version builder to enable adding downloads to a particular version?

`CrateBuilder::version` can also take a `VersionBuilder` rather than just a version string, so the test setup could get rid of these highlighted lines and do something like this instead:

```
let _krate = CrateBuilder::new("foo_versions", user.id)
            .version(VersionBuilder::new("0.5.1").recent_downloads(3))
            .version(VersionBuilder::new("1.0.0").recent_downloads(2))
            .expect_build(conn);
```

I'm also wondering if we should be testing that this endpoint doesn't include non-recent download stats, so we'd need a way to set older downloads. The crate builder isn't a great example for this as it kind of cheats. `VersionDownload` also doesn't have a nice function for setting downloads for a particular date because usually we only want to set downloads for today. The [populate](https://github.com/rust-lang/crates.io/blob/master/src/bin/populate.rs) utility script for setting up test data has an example of doing this, though.

What do you think? Does that make sense?

Reply:
I've updated the tests to try and add clarity to what is going on. The example you gave was close to working, what is the correct api is:

```rust
let _krate = CrateBuilder::new("foo_versions", user.id)
            .version(VersionBuilder::new("0.5.1")).recent_downloads(3)
            .version(VersionBuilder::new("1.0.0")).recent_downloads(2)
            .expect_build(conn);
```

however this only adds downloads to the most recently added downloaded version (i.e. 1.0.0). I had to use diesel to insert rows into the `version_downloads` table manaully, though I tried to add clarity.

> I'm also wondering if we should be testing that this endpoint doesn't include non-recent download stats, so we'd need a way to set older downloads. The crate builder isn't a great example for this as it kind of cheats. VersionDownload also doesn't have a nice function for setting downloads for a particular date because usually we only want to set downloads for today. The populate utility script for setting up test data has an example of doing this, though.

I also added this to the test, so now it sets up two versions within 90 days and two versions older than 90 days. I _also_ added some checking that the versions come back sorted, which should make the tests stable and reliable.